Add validation to disk's name in CreateVMWizard Storage step

bamsalem opened the pull request:
This PR adds validations to disk name such as in NIC name in Network Step.
The rules for the validation taken from [the documentation](https://access.redhat.com/documentation/en-us/red_hat_virtualization/4.4/html/administration_guide/sect-virtual_disk_tasks).
In case of error it will show error message, mark the field in red and not let user save the edited disk and move to the next step:

![image](https://user-images.githubusercontent.com/64131213/102999450-ae9a6480-4531-11eb-9b1b-004a65d1a4cb.png)

If you have any questions please feel free to ask
Fixes: #1329

ovirt-infra commented:
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

sgratch commented:
ci add to whitelist

sgratch commented on src/intl/messages.js:
```diff
@@ -167,6 +167,7 @@ export const messages: { [messageId: string]: MessageType } = {
   createVmWizardStepTitleNetwork: 'Networking',
   createVmWizardStepTitleReview: 'Review',
   createVmWizardStepTitleStorage: 'Storage',
+  createVmWizardStorageDiskNameRules: 'Disk name must be 40 or less alphanumeric characters or "-_."',
```
This message is not accurate, please check that you can create a disk name longer than 40 chars. Please rephrase it to be the same as on webadmin:
Disk name can contain only alphanumeric, '.', '_' or '-' characters.

bamsalem replied:
![image](https://user-images.githubusercontent.com/64131213/103477451-b4ad0280-4dc7-11eb-9bc4-2c9a8e0f085a.png)

sgratch commented on src/components/utils/validation.js:
```diff
@@ -88,7 +88,7 @@ export function isVmNameValid (nameCandidate: string): boolean {
 
 export function isDiskNameValid (nameCandidate: string): boolean {
-  return isVmNameValid(nameCandidate)
+  return /^[a-zA-Z0-9_\-\\.]{1,40}$/.test(nameCandidate)
```
Please update  this length validation "{1,40}" to a much longer length (e.g. 64 chars)

bamsalem replied:
@sgratch Are you sure?
I stuck to the [documentation](https://access.redhat.com/documentation/en-us/red_hat_virtualization/4.3/html/administration_guide/sect-virtual_disk_tasks#Creating_a_Virtual_Disk)

sgratch replied:
> @sgratch Are you sure?
> I stuck to the [documentation](https://access.redhat.com/documentation/en-us/red_hat_virtualization/4.3/html/administration_guide/sect-virtual_disk_tasks#Creating_a_Virtual_Disk)

When comparing webadmin code/behaviour to the documentation, usually the webadmin behaviour is the the enforcing rule (and probably the doc is not up to date) :). 
Specifically here, the limit is 255 chars:
https://github.com/oVirt/ovirt-engine/blob/e48484ccb3c1c8fcaed81023b234295a109ebd8b/frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/DisksAllocationModel.java#L486-L490

bamsalem replied:
OK, I'll check webAdmin code next time :)

sgratch reviewed (dismissed):
Please see inline comments.

In addition, there is a problem of saving a new disk after an error is displayed:
1. create  a new disk with invalid name - validation failed and an inline error is displayed.
2. fix the disk name to become a valid one.

Expected: saving the disk is possible.
Actual behaviour: no error is displayed but you can't save the disk

sgratch commented on src/components/utils/validation.js:
```diff
@@ -88,7 +88,7 @@ export function isVmNameValid (nameCandidate: string): boolean {
 
 export function isDiskNameValid (nameCandidate: string): boolean {
-  return isVmNameValid(nameCandidate)
+  return /^[a-zA-Z0-9_\-\\.]{1,64}$/.test(nameCandidate)
```
Why limiting to 64 chars while on webadmin and backend we now know that the limitation is for 255 chars? It's better to use the same limitation

sgratch commented on src/components/CreateVmWizard/steps/Storage.js:
```diff
@@ -503,21 +517,32 @@ class Storage extends React.Component {
+        break
+      case 'name':
+        editingErrors.diskInvalidName = !isDiskNameValid(value)
+        console.log('editingErrors', editingErrors)
```
Don't you need to delete this console.log line? I can't see the benefit of leaving it.

sgratch reviewed (commented):
A small inline comment.

rszwajko commented on src/components/utils/validation.js:
```diff
@@ -88,7 +88,7 @@ export function isVmNameValid (nameCandidate: string): boolean {
 
 export function isDiskNameValid (nameCandidate: string): boolean {
-  return isVmNameValid(nameCandidate)
+  return /^[a-zA-Z0-9_\-\\.]{1,255}$/.test(nameCandidate)
```
Please add an unit test  - previously `isVmNameValid()` was unit tested.
@sgratch  @sjd78 
Validation is now more restrictive. However `isDiskNameValid` is also used to validate alias name in `DiskImageEditor`. Can we break some use case here?

bamsalem replied:
I added unit test :)

sgratch replied:
@bamsalem @rszwajko regarding using isDiskNameValid in both create/edit modes:
we can't break any use case since isVmNameValid and isDiskNameValid actually share the same validation logic except length limitation, so now validating of disk alias name in DiskImageEditor is correct.
But two issues just bumped up:
-  is there any reason for using VM_NAME_REGEX.test(nameCandidate) within isVmNameValid function and  /^[a-zA-Z0-9_\-\\.]{1,255}$/.test(nameCandidate) within isDiskNameValid function? We can check the length limitation separately and share the same regex...

- for both DiskImageEditor and Storage.js we use the same validation but display a different error message (pleaseEnterValidDiskName and createVmWizardStorageDiskNameRules). I suggest using the same message.

rszwajko replied:
@sgratch  my understanding is that VM_NAME_REGEXP is **less** strict - check [UNICODE_NON_ASCII_LETTERS](https://github.com/oVirt/ovirt-web-ui/blob/a41b92f1e340480129c2cd5f6c9da58066ef96ce/src/components/utils/validation.js#L64)

bamsalem replied:
@sgratch
I fixed the validations now we are using only the /^[a-zA-Z0-9_-\.]{1,255}$/.test(nameCandidate) validation.

Regarding the message we display different messages because the message in Create VM wizard must be shorter than DiskImageEditor, that more descriptive:

Long message:
![image](https://user-images.githubusercontent.com/64131213/105989734-35e77480-60aa-11eb-8832-41d87b638372.png)

Short message:
![image](https://user-images.githubusercontent.com/64131213/105990065-a7272780-60aa-11eb-9aa8-059762a719c2.png)

If you still think we should use the same message, WDYT? which one is preferred?

bamsalem commented on src/components/CreateVmWizard/steps/Storage.js:
```diff
@@ -503,21 +517,32 @@ class Storage extends React.Component {
+
+        const { minDiskSizeInGiB, maxDiskSizeInGiB } = this.props
+        if (value >= minDiskSizeInGiB && value <= maxDiskSizeInGiB) {
+          delete editingRow.invalidSizeValue
```
it was part of another fix, I'll change that

bamsalem commented on src/components/CreateVmWizard/steps/Storage.js:
```diff
@@ -105,6 +107,9 @@ class Storage extends React.Component {
     this.isEditingMode = this.isEditingMode.bind(this)
 
     this.state = {
+      editingErrors: {
```
`editing` is an object that includes the data of the edited/created disk, and `creating`  just holds the id of the created disk in case of editing its value is false, so the answer is no, hope I explained it well :)

Creating disk:
![creating](https://user-images.githubusercontent.com/64131213/104180510-e7817700-5415-11eb-9945-eddae415de92.png)
Editing disk:
![editing](https://user-images.githubusercontent.com/64131213/104180935-9faf1f80-5416-11eb-83d3-02aa5657ccfe.png)

sjd78 commented on src/components/utils/validation.test.js:
```diff
@@ -20,6 +20,28 @@ describe('check host names', function () {
 
+describe('check disk names', function () {
+  it('valid disk names', function () {
+    expect(isDiskNameValid('veryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryValidDiskName12')).toEqual(true)
```
```suggestion
    expect(isDiskNameValid('abcde'.repeat(51))).toEqual(true) // 255 characters
```

sjd78 commented:
@sgratch , @bamsalem , @rszwajko - The VM name regexp needs to be changed back.  VMs can have names that do not contain latin alpha characters.  In fact, the regexp is a copy from webadmin's VM name validation.

I'm not sure if disk names can contain non-latin alpha characters.  I did not test that, or lookup the validation code, yet.

rszwajko commented on src/components/utils/validation.test.js:
```diff
@@ -18,8 +18,30 @@ describe('check host names', function () {
+    expect(isDiskNameValid('a!@$b298')).toEqual(false)
+    expect(isDiskNameValid('a_name()')).toEqual(false)
+    expect(isDiskNameValid('Кирилиця')).toEqual(false)
+    expect(isDiskNameValid('veryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryverylongDiskName')).toEqual(false)
```
How about this: `'a'.repeat(256)`? we want to check exactly one above the limit

sjd78 replied:
> How about this: `'a'.repeat(256)`? we want to check exactly one above the limit

Even better

rszwajko commented on src/components/utils/validation.js:
```diff
@@ -1,97 +1,42 @@
 // @flow
 
+const charactersRegExpTest = (nameCandidate: string) => /^[a-zA-Z0-9_\-\\.]*$/.test(nameCandidate)
```
@sgratch  Actually I'm not sure if we should throw away UTF chars. Check  I18NNameValidation which is used DiskAllocationModel - it [defines letters as all chars including UTF](https://github.com/oVirt/ovirt-engine/blob/e48484ccb3c1c8fcaed81023b234295a109ebd8b/frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/validation/I18NNameValidation.java#L11)

sgratch commented:
@bamsalem @sjd78 @rszwajko 

> is there any reason for using VM_NAME_REGEX.test(nameCandidate) within isVmNameValid function and /^[a-zA-Z0-9_-.]{1,255}$/.test(nameCandidate) within isDiskNameValid function? We can check the length limitation separately and share the same regex...

You are all right, based on the fact that disk alias string is auto generated from the vm name ({vmName}_Disk1) then we should support UTF for both disk name and vm name.

And by checking the code: 
disk: https://github.com/oVirt/ovirt-engine/blob/e48484ccb3c1c8fcaed81023b234295a109ebd8b/frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/DisksAllocationModel.java#L486-L490

vm: https://github.com/oVirt/ovirt-engine/blob/9059df4d26f5d327c87495ee9e19cb1876ad92b7/frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/vms/VmModelBehaviorBase.java#L1665-L1667

Both use _I18NNameValidation_().

So in our web-ui code we should use VM_NAME_REGEX.test(nameCandidate) for both validations


> Regarding the message we display different messages because the message in Create VM wizard must be shorter than DiskImageEditor, that more descriptive

I'm totally ok with having the shorter message version for both dialogs. They both cover the same info and it's better to maintain one message instead of two (if validation login will be ever changed).

bamsalem commented:
I changed it to use VM_NAME_REGEX again in diskName and vmName.
I changed the code according to your review @sgratch @sjd78 @rszwajko...


> I'm totally ok with having the shorter message version for both dialogs. They both cover the same info and it's better to maintain one message instead of two (if validation login will be ever changed).

Now the same message displayed in CreateVm and editVm:
![image](https://user-images.githubusercontent.com/64131213/106384382-41e97400-63d3-11eb-99a1-64bd2c22e8eb.png)

sgratch commented:
@bamsalem please rebase

sjd78 reviewed (approved):
Component code and error messages look good.
Validation checks and tests look good as well.

bamsalem commented:
@sgratch 
rebased.

isaranova commented:
LGTM
